testcontainers/src/clients/http.rs

```diff
@@ -33,46 +30,112 @@ struct Client {
     created_networks: RwLock<Vec<String>>,
 }
 
-impl fmt::Debug for Http {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Http").finish()
+impl Default for Client {
+    fn default() -> Self {
+        Client {
+            command: env::command::<env::Os>().unwrap_or_default(),
+            bollard: Docker::connect_with_http_defaults().unwrap(),
+            created_networks: RwLock::new(Vec::new()),
+        }
     }
 }
 
-impl Default for Http {
-    fn default() -> Self {
-        Self::new()
+impl Client {
+    async fn create_network_if_not_exists(&self, network: &str) -> bool {
+        if !network_exists(&self.bollard, network).await {
+            self.bollard
+                .create_network(CreateNetworkOptions {
+                    name: network.to_owned(),
+                    ..Default::default()
+                })
+                .await
+                .unwrap();
+
+            return true;
+        }
+
+        false
+    }
+
+    async fn create_container(
+        &self,
+        options: Option<CreateContainerOptions<String>>,
+        config: Config<String>,
+    ) -> Result<ContainerCreateResponse, bollard::errors::Error> {
+        self.bollard.create_container(options, config).await
+    }
+
+    fn logs(&self, container_id: String, options: LogsOptions<String>) -> LogStreamAsync<'_> {
+        let stream = self
+            .bollard
+            .logs(&container_id, Some(options))
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))
+            .map(|chunk| {
+                let bytes = chunk?.into_bytes();
+                let str = std::str::from_utf8(bytes.as_ref())
+                    .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+                Ok(String::from(str))
+            })
+            .boxed();
+
+        LogStreamAsync::new(stream)
+    }
+}
+
+async fn network_exists(client: &Docker, network: &str) -> bool {
+    let networks = client.list_networks::<String>(None).await.unwrap();
+    networks
+        .iter()
+        .any(|i| matches!(&i.name, Some(name) if name == network))
+}
+
+impl Drop for Client {
+    fn drop(&mut self) {
+        match self.command {
+            env::Command::Remove => {
+                let guard = self.created_networks.read().expect("failed to lock RwLock");
+                for network in guard.iter() {
+                    block_on(async { self.bollard.remove_network(network).await.unwrap() });
+                }
+            }
+            env::Command::Keep => {}
+        }
```

Comment:
This is something we need to talk I believe. I recently learned that static variables don't actually call `Drop` so we might have to move this logic to `WatchDog` or find another way to create global static variable. 

Arc could be useful but I am not sure how to create and Arc that behaves like static variable.

Reply:
Yeah a `static` variable will only be dropped at the end of a program.

Rust executes each test in its own thread so we could look into using a thread-local variable for this. That would clean up a docker client at the end of each thread, i.e. at the end of each test and each test would gets its own docker client.

Reply:
Now the tests are updated there are 3 failing tests and all of them about related to "networks" which were suppose to be cleaned up but it's not working due to the issue I mentioned.

I see 3 options here;
- The method you mentioned - (I am not very familiar with rust in that regard so if you let me know how this could be achieved I can take a look)

- We can integrated with [ryuk](https://github.com/testcontainers/moby-ryuk) which would handle all kind of clean-up scenarios also align with other testcontainers projects. (I am not %100 sure but this would mean that some test cases can't be run in parallel) This approach is similar to watchdog feature we have in this repo but ryuk is client agnostic.
 
- For this specific use case, we can keep track of number of containers which uses the given network and remove at when the last container is dropped.

```
// assume this is global variable 
let networks = Mutex<Map<String,i32>> // network_name -> number_of_containers_using_the_network

// inside start function
insert or increase the number of reference to given network key in global map

// inside impl drop for container
decrease the number of reference to given network. If number of reference is 0 after this operation remove the network as well.
```

I personally prefer option two so we can align with other implementations and also I like the ryuk approach in terms of design but I would like to hear what you think.

Reply:
Essentially, what we need to achieve is that a client is only alive for as long as any of its containers are alive. Thus I'd suggest the following:

- Store a `Mutex<Option<Weak<Client>>>` in the static variable
- When you try to get a reference to the client:
	- Grab a lock to the mutex
	- If the value is `None`:
		- Make a new client
		- Wrap it in an `Arc`
		- Make a copy
		- Downgrade the copy to a `Weak`
		- Store the `Weak` inside the `Option`
		- Release the lock
		- Return the `Arc` you still have around
	- If the value is `Some`:
		- Attempt to upgrade to an `Arc`
			- If successful, return
			- If failed, do the same thing as for the above `None` case

I believe this should achieve the following:

- As long is there is at least 1 `Container` around, we can grab a reference to it and make more containers.
- If all containers are dropped, the last strong reference to the `Client` is also dropped which means `Drop` is being called on `Client`

I see integration with `ryuk` as a separate feature.